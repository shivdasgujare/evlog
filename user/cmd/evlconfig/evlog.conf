#
# Configuration values for evlogd daemon. 
# Use 'evlconfig' command to update these values. If this file is edited 
# manually, the updated value will not be used by the evlogd daemon. So,
# the evlogd daemon has to be restarted. Hence, use 'evlconfig' command
# to send this values to evlogd daemon.
# Console display level - OFF=111
#
Discard Dups: on
Discard Interval: 3
Discard Count: 100
Event Screen: nofilter
Console display level: 111
Lookback Size: 3
