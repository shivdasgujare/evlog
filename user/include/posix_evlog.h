/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2001
 *
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */

#define LOG_LOGMGMT (12<<3) /* this facility defined by POSIX, but not syslog */
#ifndef _POSIX_EVLOG_H_
#define _POSIX_EVLOG_H_

#include <sys/types.h>
/* For sigevent */
#include <signal.h>
/* For facility (e.g., LOG_KERN) and severity (e.g., LOG_EMERG) codes */
#include <syslog.h>
/* For va_list */
#include <stdarg.h>

#ifdef __cplusplus
extern "C" {
#endif	/* __cplusplus */

#ifdef _POSIX_THREADS
#include <pthread.h>
#endif
#include <linux/evl_log.h>

/* Treat a log descriptor like a file descriptor for now. */
typedef struct _log_description *posix_logd_t;

/* Member selectors for posix_log_entry, for use by posix_log_memtostr  */
#define POSIX_LOG_ENTRY_DATA	0	/* variable-length data */
#define POSIX_LOG_ENTRY_RECID	1
#define POSIX_LOG_ENTRY_SIZE	2
#define POSIX_LOG_ENTRY_FORMAT	3
#define POSIX_LOG_ENTRY_EVENT_TYPE	4
#define POSIX_LOG_ENTRY_FACILITY	5
#define POSIX_LOG_ENTRY_SEVERITY	6
#define POSIX_LOG_ENTRY_UID		7
#define POSIX_LOG_ENTRY_GID		8
#define POSIX_LOG_ENTRY_PID		9
#define POSIX_LOG_ENTRY_PGRP	10
#define POSIX_LOG_ENTRY_TIME	11
#define POSIX_LOG_ENTRY_FLAGS	12
#define POSIX_LOG_ENTRY_THREAD	13
#define POSIX_LOG_ENTRY_PROCESSOR	14

/* Query purpose flags */
#define POSIX_LOG_PRPS_NOTIFY	0x1
#define POSIX_LOG_PRPS_GENERAL	0x2
#define POSIX_LOG_PRPS_SEEK	(POSIX_LOG_PRPS_GENERAL)
#define EVL_PRPS_TEMPLATE	0x4
#define EVL_PRPS_RESTRICTED	0x8
#define POSIX_LOG_PRPS_MASK	0xf

struct node;

typedef struct _posix_log_query {
	int		qu_purpose;	/* NOTIFY, SEEK, etc. */
	char		*qu_expr;	/* Saved copy of the query expr */
	struct node	*qu_tree;	/* Parse tree */
	struct evl_nonStdAtts	*qu_nonStdAtts;	/* Non-standard attributes. */
} posix_log_query_t;


#define POSIX_LOG_ONCE_ONLY		0x01
#define POSIX_LOG_SEND_RECID		0x02
#define POSIX_LOG_SEND_SIGVAL		0x04
#define POSIX_LOG_NFY_MASK		0x07

#define POSIX_LOG_NFY_DISABLED		0x08
#define POSIX_LOG_ACTION_PERSIST	0x16

#define POSIX_LOG_NFY_PENDING_REMOVE	0x512

#define SI_EVLOG	SI_QUEUE
typedef size_t posix_log_notify_t;
/*
 * These are the types of messages that can be sent between the client (a
 * process that has called posix_log_notify_add_ex()) and the server (the
 * notification daemon).
 */
typedef enum _nfyMsgType {
	nmtNewRequest,			/* new notification request */
	nmtGetRequestStatus,	/* get status of notification request */
	nmtRmRequest,			/* remove notification request */
	nmtRequestStatus		/* response to client requests */
} notifyMsgType_t;

/*
 * The status of a notification request.
 */
typedef enum _nfyRqStatus {
	nrsEnabled,		/* Request has been accepted by server. */
	nrsDisabled,	/* Once-only request has been satisfied */
					/* by the server, but not removed by the */
					/* client. */
	nrsRemoved,		/* Request has been removed at client's request. */
	nrsNoRequest	/* Request not found by server */
} nfyRqStatus_t;

/*
 * The header of a message between client and server.
 */
typedef struct _nfyMsgHdr {
	int	nhMsgType;
	posix_log_notify_t	nhReqHandle;
	int	nhStatus;
} nfyMsgHdr_t;

typedef struct _nfyNewRqMsg {
	nfyMsgHdr_t	nwHeader;
	size_t		nwQueryLength;
	struct sigevent	nwSigevent;
	int		nwFlags;
} nfyNewRqMsg_t;

#define EVL_NFY_MSGMAXSIZE (sizeof(nfyNewRqMsg_t))

/* Log-management event-type codes */
#define POSIX_LOG_MGMT_TIMEMARK	1
#define POSIX_LOG_MGMT_STARTMAINT	2
#define POSIX_LOG_MGMT_ENDMAINT	3
#define POSIX_LOG_MGMT_STOPLOGGING	4
#define POSIX_LOG_MGMT_STARTLOGGING	5

/* Per-process maximum number of concurrent open log descriptors */
#define POSIX_LOG_OPEN_MAX	100

/* Per-process max number of concurrently registered notification requests */
#define POSIX_LOG_NOTIFY_MAX	100

/* Seek directions */
#define POSIX_LOG_SEEK_START	1
#define POSIX_LOG_SEEK_END		2
#define POSIX_LOG_SEEK_FORWARD	3
#define POSIX_LOG_SEEK_BACKWARD	4
#define POSIX_LOG_SEEK_FIRST	5
#define POSIX_LOG_SEEK_LAST		6

/* Maximum length for a string returned by posix_log_memtostr */
#define POSIX_LOG_MEMSTR_MAXLEN	128

/* Log an event */
extern int posix_log_write(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity,
	const void *buf, size_t len, int format, unsigned int flags);

extern int posix_log_vprintf(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity, unsigned int flags,
	const char *format, va_list args);

extern int posix_log_printf(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity, unsigned int flags,
	const char *format, ...);

extern int posix_log_vprintb(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity, unsigned int flags,
	const char *format, va_list args);

extern int posix_log_printb(posix_log_facility_t facility,
	int event_type, posix_log_severity_t severity, unsigned int flags,
	const char *format, ...);

/* Open an event log for read access */
extern int posix_log_open(posix_logd_t *logdes, const char *path);

/* Read from event log */
extern int posix_log_read(posix_logd_t logdes, struct posix_log_entry *entry,
	void *log_buf, size_t log_len); 

/* Register for notification about new events */
extern int posix_log_notify_add(const posix_log_query_t *query,
	const struct sigevent *notification, int flags,
	posix_log_notify_t *nfyhandle);

extern int posix_log_sigval_getrecid(const union sigval sval,
	posix_log_recid_t *recid);

extern int posix_log_siginfo_getrecid(const siginfo_t *info,
        void *context, posix_log_recid_t *recid);
        
extern int posix_log_notify_get(posix_log_notify_t nfyhandle,
	struct sigevent *notification, int *flags,
	char *qsbuf, size_t qslen, size_t *reqlen);

/* Remove notification request */
extern int posix_log_notify_remove(posix_log_notify_t nfyhandle);

/* Close event log */
extern int posix_log_close(posix_logd_t logdes);

/* Reposition the read pointer */
extern int posix_log_seek(posix_logd_t logdes, const posix_log_query_t *query,
	int direction); 

/* Compare severities */
extern int posix_log_severity_compare(int *order, posix_log_severity_t s1,
	posix_log_severity_t s2); 

/* Create log query */
extern int posix_log_query_create(const char *query_string, int purpose,
	posix_log_query_t *query, char *errbuf, size_t errlen);

extern int posix_log_query_get(const posix_log_query_t *query, int *purpose,
	char *qsbuf, size_t qslen, size_t *reqlen);

/* Destroy log query */
extern int posix_log_query_destroy(posix_log_query_t *query);

/* Obtain the string equivalent of an event-record member */
extern int posix_log_memtostr(int member_selector,
	const struct posix_log_entry *entry,
	char *buf, size_t buflen);

/* Convert string to facility or facility to string */
extern int posix_log_factostr(posix_log_facility_t fac, char *buf,
	size_t buflen);

extern int posix_log_strtofac(const char *str, posix_log_facility_t *fac);

#ifdef __cplusplus
}
#endif 	/* __cplusplus */
#endif	/* _POSIX_EVLOG_H_ */
