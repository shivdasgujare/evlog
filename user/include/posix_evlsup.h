/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2001
 *
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */
#include <sys/un.h>
#ifndef _POSIX_EVLSUP_H_
#define _POSIX_EVLSUP_H_ 1

#ifdef __cplusplus
extern "C" {
#endif	/* __cplusplus */

typedef size_t HANDLE;

/* Name-value pair */
struct _evlNvPair {
	int	nv_value;
	char	*nv_name;
};

/* Values for faAccessFlags */
#define EVL_FACC_PRIVATE 0x1	/* facility's event recs are read-protected */
#define EVL_FACC_KERN 	 0x2

/* An entry from the facility registry */
typedef struct __evlFacility {
	posix_log_facility_t	faCode;
	char					*faName;
	int						faAccessFlags;
	char 					*faFilter;
	posix_log_query_t		*faQuery;
} _evlFacility;

/* Internal view of the info in /etc/evlog.d/facility_registry */
typedef struct __evlFacilityRegistry {
	int			frNFacilities;
	_evlFacility		*frHash;	/* hash table */
	int			frHashSize;	/* # slots in hash table */
} _evlFacilityRegistry;

extern void _evlGetNameByValue(const struct _evlNvPair table[], int value,
	char *name, size_t size, const char *dflt);
extern int _evlGetValueByName(const struct _evlNvPair table[], const char *name,
	int dflt);
extern int _evlGetValueByCIName(const struct _evlNvPair table[],
	const char *name, int dflt);
extern int _evlCIStrcmp(const char *s1, const char *s2);

extern struct _evlNvPair _evlLinuxFacilities[];
extern struct _evlNvPair _evlPosixFacilities[];
extern struct _evlNvPair _evlSeverities[];
extern struct _evlNvPair _evlAttributes[];
extern struct _evlNvPair _evlFormats[];
extern struct _evlNvPair _evlMgmtEventTypes[];

#define LOG_EVLOG_DIR EVLOG_STATE_DIR
#define LOG_EVLOG_CONF_DIR "/etc/evlog.d"
#define LOG_CURLOG_PATH LOG_EVLOG_DIR "/eventlog"
#define LOG_PRIVATE_PATH LOG_EVLOG_DIR "/privatelog"

#define EVL_INVALID_FACILITY ((posix_log_facility_t)-1)
#define EVL_INVALID_EVENT_TYPE (-1)

/*
 * The "age" pseudo-attribute -- must be distinct from POSIX_LOG_ENTRY_* in
 * posix_evlog.h.
 */
#define EVL_ENTRY_AGE 20
#define EVL_ENTRY_HOST 21

/* from posix2.c */
extern int _evlWrite(struct posix_log_entry *entry, const void *buf);
extern int _evlOpen(posix_logd_t *logdes, const char *path);
extern int _evlValidateLogdes(posix_logd_t logdes, int checkGeneration);
extern int _evlRead(posix_logd_t logdes, struct posix_log_entry *entry,
	void *buf, size_t bufLen);
extern int _evlClose(posix_logd_t fd);
extern int _evlRewind(posix_logd_t logdes);
extern int _evlSeekEnd(posix_logd_t logdes);
extern int _evlSeekFwd(posix_logd_t logdes, const posix_log_query_t *query);
extern int _evlSeekBkwd(posix_logd_t logdes, const posix_log_query_t *query);
extern int _evlFdWrite(int fd, const struct posix_log_entry *entry,
	const void *buf);
extern int _evlWriteLogHeader(int fd);
extern int _evlValidateQuery(posix_log_query_t *query, int nullOk);

/* facility registry functions */
extern void _evlLockFacReg(void);
extern void _evlUnlockFacReg(void);
extern int _evlBlockSignals(sigset_t *oldset);
extern void _evlRestoreSignals(sigset_t *oldset);
extern _evlFacilityRegistry *_evlReadFacilities(const char *facRegPath);
extern posix_log_facility_t _evlGetFacilityCodeByCIName(const char *name);
extern char *_evlGetFacilityName(posix_log_facility_t facNum, char *buf);
extern int _evlGetFacilityAccess(posix_log_facility_t facNum);
extern posix_log_query_t * _evlGetFacilityAccessQuery(posix_log_facility_t facNum, int * acc);
extern struct _evlNvPair *_evlSnapshotFacilities();
extern int _evlAddFacilityToRegistry(const char *fname, unsigned int flags,
	const char *filter);
extern unsigned int _evlCrc32(const unsigned char *data, int len);

/* posix1.c */
extern int _evlGetProcId();
extern int isAccess(int sd);
extern int _establishNonBlkConnection(const char * socketpath, struct sockaddr_un *sa, int nsec);
extern int _evlWriteEx(struct posix_log_entry *hdr, const char *buf);
extern int _evlEvaluateQuery(const posix_log_query_t *query,
	const struct posix_log_entry *entry, const void *buf);

#ifdef __cplusplus
}
#endif	/* __cplusplus */
#endif 	/* _POSIX_EVLSUP_H_ */
