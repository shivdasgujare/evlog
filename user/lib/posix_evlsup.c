/*
 * Linux Event Logging for the Enterprise
 * Copyright (c) International Business Machines Corp., 2001
 *
 *
 *  This library is free software; you can redistribute it and/or
 *  modify it under the terms of the GNU Lesser General Public
 *  License as published by the Free Software Foundation; either
 *  version 2.1 of the License, or (at your option) any later version.
 *
 *  This library is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 *  Lesser General Public License for more details.
 *
 *  You should have received a copy of the GNU Lesser General Public
 *  License along with this library; if not, write to the Free Software
 *  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
 *
 *  Please send e-mail to lkessler@users.sourceforge.net if you have
 *  questions or comments.
 *
 *  Project Website:  http://evlog.sourceforge.net/
 *
 */
#include <stdio.h>
#include <unistd.h>
#include <string.h>
#include <ctype.h>
#include <sys/syslog.h>
#include <sys/stat.h>
#include <sys/types.h>
#include <fcntl.h>
#include <assert.h>
#include <stdlib.h>

#include "config.h"

#include "posix_evlog.h"
#include "posix_evlsup.h"
#include "evlog.h"

/*
 * These are the facilities defined by Linux's sys/syslog.h.
 * The facility registry must always contain at least these.
 */
struct _evlNvPair _evlLinuxFacilities[] = {
	{ LOG_AUTH,	"AUTH" },
	{ LOG_CRON,	"CRON" },
	{ LOG_DAEMON,	"DAEMON" },
	{ LOG_KERN,	"KERN" },
	{ LOG_LPR,	"LPR" },
	{ LOG_MAIL,	"MAIL" },
	{ LOG_NEWS,	"NEWS" },
	{ LOG_SYSLOG,	"SYSLOG" },
	{ LOG_USER,	"USER" },
	{ LOG_LOGMGMT,	"LOGMGMT" },

	{ LOG_UUCP,	"UUCP" },
	{ LOG_AUTHPRIV,	"AUTHPRIV" },
	{ LOG_FTP,	"FTP" },

	{ LOG_LOCAL0,	"LOCAL0" },
	{ LOG_LOCAL1,	"LOCAL1" },
	{ LOG_LOCAL2,	"LOCAL2" },
	{ LOG_LOCAL3,	"LOCAL3" },
	{ LOG_LOCAL4,	"LOCAL4" },
	{ LOG_LOCAL5,	"LOCAL5" },
	{ LOG_LOCAL6,	"LOCAL6" },
	{ LOG_LOCAL7,	"LOCAL7" },
	{ 0, 0 }
};

/*
 * These are the facilities defined by POSIX 1003.25.
 */
struct _evlNvPair _evlPosixFacilities[] = {
	{ LOG_AUTH,	"AUTH" },
	{ LOG_CRON,	"CRON" },
	{ LOG_DAEMON,	"DAEMON" },
	{ LOG_KERN,	"KERN" },
	{ LOG_LPR,	"LPR" },
	{ LOG_MAIL,	"MAIL" },
	{ LOG_NEWS,	"NEWS" },
	{ LOG_SYSLOG,	"SYSLOG" },
	{ LOG_USER,	"USER" },
	{ LOG_LOGMGMT,	"LOGMGMT" },
	{ LOG_UUCP,	"UUCP" },
	{ LOG_LOCAL0,	"LOCAL0" },
	{ LOG_LOCAL1,	"LOCAL1" },
	{ LOG_LOCAL2,	"LOCAL2" },
	{ LOG_LOCAL3,	"LOCAL3" },
	{ LOG_LOCAL4,	"LOCAL4" },
	{ LOG_LOCAL5,	"LOCAL5" },
	{ LOG_LOCAL6,	"LOCAL6" },
	{ LOG_LOCAL7,	"LOCAL7" },
	{ 0, 0 }
};

struct _evlNvPair _evlSeverities[] = {
	{ LOG_EMERG,	"EMERG" },
	{ LOG_ALERT,	"ALERT" },
	{ LOG_CRIT,	"CRIT" },
	{ LOG_ERR,	"ERR" },
	{ LOG_WARNING,	"WARNING" },
	{ LOG_NOTICE,	"NOTICE" },
	{ LOG_INFO,	"INFO" },
	{ LOG_DEBUG,	"DEBUG" },
	{ 0, 0 }
};

struct _evlNvPair _evlAttributes[] = {
	{ POSIX_LOG_ENTRY_RECID,	"recid" },
	{ POSIX_LOG_ENTRY_SIZE,	"size" },
	{ POSIX_LOG_ENTRY_FORMAT,	"format" },
	{ POSIX_LOG_ENTRY_EVENT_TYPE,	"event_type" },
	{ POSIX_LOG_ENTRY_FACILITY,	"facility" },
	{ POSIX_LOG_ENTRY_SEVERITY,	"severity" },
	{ POSIX_LOG_ENTRY_UID,	"uid" },
	{ POSIX_LOG_ENTRY_GID,	"gid" },
	{ POSIX_LOG_ENTRY_PID,	"pid" },
	{ POSIX_LOG_ENTRY_PGRP,	"pgrp" },
	{ POSIX_LOG_ENTRY_TIME,	"time" },
	{ POSIX_LOG_ENTRY_FLAGS,	"flags" },
	{ POSIX_LOG_ENTRY_THREAD,	"thread" },
	{ POSIX_LOG_ENTRY_PROCESSOR,	"processor" },
	{ 0, 0 }
};

struct _evlNvPair _evlFormats[] = {
	{ POSIX_LOG_NODATA,	"NODATA" },
	{ POSIX_LOG_BINARY,	"BINARY" },
	{ POSIX_LOG_STRING,	"STRING" },
	{ POSIX_LOG_PRINTF,	"PRINTF" },
	{ 0, 0 }
};

struct _evlNvPair _evlMgmtEventTypes[] = {
	{ POSIX_LOG_MGMT_TIMEMARK,		"MGMT_TIMEMARK" },
	{ POSIX_LOG_MGMT_STARTMAINT,	"MGMT_STARTMAINT" },
	{ POSIX_LOG_MGMT_ENDMAINT,		"MGMT_ENDMAINT" },
	{ POSIX_LOG_MGMT_STOPLOGGING,	"MGMT_STOPLOGGING" },
	{ POSIX_LOG_MGMT_STARTLOGGING,	"MGMT_STARTLOGGING" },
	{ 0, 0 }
};

void
_evlGetNameByValue(const struct _evlNvPair table[], int value, char *name, size_t size,
	const char *dflt)
{
	const struct _evlNvPair *nv;
	for (nv = table; nv->nv_name; nv++) {
		if (nv->nv_value == value) {
			(void) strcpy(name, nv->nv_name);
			return;
		}
	}

	/* No match */
	if (dflt) {
		(void) strcpy(name, dflt);
	} else {
		snprintf(name, size, "%d", value);
	}
}

/* Look up value by name, case-sensitive */
int
_evlGetValueByName(const struct _evlNvPair table[], const char *name, int dflt)
{
	const struct _evlNvPair *nv;
	for (nv = table; nv->nv_name; nv++) {
		if (!strcmp(nv->nv_name, name)) {
			return nv->nv_value;
		}
	}

	/* No match */
	return dflt;
}

/* Like strcmp, but case-insensitive */
int
_evlCIStrcmp(const char *s1, const char *s2)
{
	while (*s1 && toupper(*s1) == toupper(*s2)) {
		s1++;
		s2++;
	}
	return toupper(*s1) - toupper(*s2);
}

/* Look up value by name, case-insensitive */
int
_evlGetValueByCIName(const struct _evlNvPair table[], const char *name, int dflt)
{
	const struct _evlNvPair *nv;
	for (nv = table; nv->nv_name; nv++) {
		if (!_evlCIStrcmp(nv->nv_name, name)) {
			return nv->nv_value;
		}
	}

	/* No match */
	return dflt;
}

/* Return the proc id */
int 
_evlGetProcId()
{
	int fd, ret;
	char buf[1024];
	char *last_token;
	size_t r;
	
	if ((fd = open("/proc/self/stat", O_RDONLY)) == -1) {
		fprintf(stderr, "Can't open stat\n");
		exit(1);
	}

	if ((r = read(fd, buf, 1024)) == -1) {
		close(fd);
		fprintf(stderr, "Read stat failed\n");
		exit(1);
	}
	buf[r]='\0';
	close(fd);
	/* The last token contains the processor id */

	last_token = (char *) strrchr(buf, ' ');
	assert(last_token != NULL);	
	return (int) strtol(last_token, (char **)NULL, 10);
}
