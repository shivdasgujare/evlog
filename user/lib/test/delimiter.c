#include <stdlib.h>
#include <stdio.h>
#include <string.h>
#include <errno.h>
#include "posix_evlog.h"
#include "posix_evlsup.h"
#include "evlog.h"

typedef struct tester {
	int one;
	int two;
} struct_t;

static void
check_status(int status, int event_type, const char *func)
{
	if (status != 0) {
		fprintf(stderr, "Error logging event type %d (0x%x):\n",
			event_type, event_type);
		errno = status;
		perror(func);
		exit(1);
	}
}

int main(int argc, char **argv)
{
	struct_t array[2];
	int status;
	array[0].one = 1;
	array[0].two = 2;
	array[1].one = 10;
	array[1].two = 20;
	
        status = posix_log_write(LOG_USER, 6000, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6000, "posix_log_write");
	
        status = posix_log_write(LOG_USER, 6001, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6001, "posix_log_write");

        status = posix_log_write(LOG_USER, 6002, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6002, "posix_log_write");

        status = posix_log_write(LOG_USER, 6003, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6003, "posix_log_write");

        status = posix_log_write(LOG_USER, 6012, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6012, "posix_log_write");

        status = posix_log_write(LOG_USER, 6013, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6013, "posix_log_write");

        status = posix_log_write(LOG_USER, 6015, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6015, "posix_log_write");

        status = posix_log_write(LOG_USER, 6016, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6016, "posix_log_write");

        status = posix_log_write(LOG_USER, 6017, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6017, "posix_log_write");

        status = evl_log_write(LOG_USER, 6021, LOG_INFO, 0, 
		"char[]", sizeof(array), &array,
		"endofdata");
	check_status(status, 6021, "evl_log_write");

        status = evl_log_write(LOG_USER, 6022, LOG_INFO, 0, 
		"char[]", sizeof(array), &array,
		"endofdata");
	check_status(status, 6022, "evl_log_write");

        status = evl_log_write(LOG_USER, 6023, LOG_INFO, 0, 
		"char[]", sizeof(array), &array,
		"endofdata");
	check_status(status, 6023, "evl_log_write");

        status = evl_log_write(LOG_USER, 6024, LOG_INFO, 0, 
		"char[]", sizeof(array), &array,
		"endofdata");
	check_status(status, 6024, "evl_log_write");

        status = evl_log_write(LOG_USER, 6025, LOG_INFO, 0, 
		"char[]", sizeof(array), &array,
		"endofdata");
	check_status(status, 6025, "evl_log_write");

        status = evl_log_write(LOG_USER, 6026, LOG_INFO, 0, 
		"char[]", sizeof(array), &array,
		"endofdata");
	check_status(status, 6026, "evl_log_write");

        status = posix_log_write(LOG_USER, 6031, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6031, "posix_log_write");

        status = posix_log_write(LOG_USER, 6032, LOG_INFO, NULL, 0, POSIX_LOG_NODATA, 0);
	check_status(status, 6032, "posix_log_write");

	status = evl_log_write(LOG_USER, 6033, LOG_INFO, 0,
		"string", "\t",
		"3*int", 10, 20, 30,
		"endofdata");
	check_status(status, 6033, "evl_log_write");

	status = evl_log_write(LOG_USER, 6034, LOG_INFO, 0,
		"string", ".",
		"3*int", 10, 20, 30,
		"endofdata");
	check_status(status, 6034, "evl_log_write");

        exit(0);
}
