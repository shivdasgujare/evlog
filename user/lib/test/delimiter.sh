#!/bin/sh
#
TEST_TEMPLATE=delimiter.out

EVLTMPLPATH=$EVL_TESTDIR/templates
export EVLTMPLPATH

if [ "$USER" = root ]
then
	/sbin/evltc $EVLTMPLPATH/user/delimiter.t
else
echo >&2 "delimiter.sh: Not running as root.  Assuming template source file"
echo >&2 "$EVLTMPLPATH/user/delimiter.t has already been compiled."
fi

lastRecid=`/sbin/evlview -t 1 -S %recid%`
./delimiter
sleep 1
/sbin/evlview -f "recid>$lastRecid && event_type > 5999 && event_type < 6100" -S '%facility% %event_type:d% %severity%:\n%data%' > delimiter.sh.out

diff delimiter.sh.out $TEST_TEMPLATE

if [ $? -eq "0" ]
then
	echo test of array delimiter		:PASSED
else
	echo test of array delimiter 		:FAILED
	exit 1
fi
rm -f delimiter.sh.out
exit 0
