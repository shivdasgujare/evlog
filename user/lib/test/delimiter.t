facility "user"; event_type 6000;
const {
        int i[3] = {10,20,30};
}
format string "%i%"
END

facility "user"; event_type 6001;
const {
        int i[3] = {10,20,30}   delimiter=", ";
}
format string "%i%"
END

facility "user"; event_type 6002;
const {
        int i[3] = {10,20,30}   "%x";
}
format string "%i%"
END

facility "user"; event_type 6003;
const {
        int i[3] = {10,20,30}   "%x" delimiter=", ";
}
format string "%i%"
END

facility "user"; event_type 6012;
const {
        int i[3] = {10,20,30}   "(%x)";
}
format string "%i%"
END

facility "user"; event_type 6013;
const {
        int i[3] = {10,20,30}   "(%x)" delimiter=", ";
}
format string "%i%"
END

facility "user"; event_type 6015;
const {
	char c[6] = {'h', 'e', 'l', 'l', 'o', '\0'}	"%s";
}
format string "%c%"
END
facility "user"; event_type 6016;
const {
	char c[6] = {'h', 'e', 'l', 'l', 'o', '\0'}	"%c";
}
format string "%c%"
END
facility "user"; event_type 6017;
const {
	char c[6] = {'h', 'e', 'l', 'l', 'o', '\0'}	"%d" delimiter = ", ";
}
format string "%c%"
END

struct struct_t;
aligned attributes {
	int one;
	int two;
}
format string "{%one%, %two%}"
END

facility "user"; event_type 6021;
attributes {
	struct struct_t str[2];
}
format
"%str%"
END
facility "user"; event_type 6022;
attributes {
	struct struct_t str[2]		delimiter=", ";
}
format
"%str%"
END
facility "user"; event_type 6023;
attributes {
	struct struct_t str[2]		"(%Z)";
}
format
"%str%"
END
facility "user"; event_type 6024;
attributes {
	struct struct_t str[2]		"(%Z)" delimiter=", ";
}
format
"%str%"
END
facility "user"; event_type 6025;
attributes {
	struct struct_t str[2]		"%Z" delimiter=", ";
}
format
"%str%"
END
facility "user"; event_type 6026;
attributes {
	struct struct_t str[2]		"%Z";
}
format
"%str%"
END

/* Tests of delimiter=attname */
struct params;
attributes {
	string delim;
}
format string "%delim%"
END
facility "user"; event_type 6031;
const {
	struct params p = { "\n" };
	int i[3] = {10,20,30}	delimiter=p.delim;
}
format string "%i%"
END
facility "user"; event_type 6032;
const {
	string delim = ":";
	int i[3] = {10,20,30}	delimiter=delim;
}
format string "%i%"
END
facility "user"; event_type 6033;
attributes {
	struct params p;
	int i[3] delimiter=p.delim;
}
format string "%i%"
END
facility "user"; event_type 6034;
attributes {
	string delim;
	int i[3] delimiter=delim;
}
format string "%i%"
END
