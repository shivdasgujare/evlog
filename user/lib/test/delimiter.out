USER 6000 INFO:
10 20 30
USER 6001 INFO:
10, 20, 30
USER 6002 INFO:
a 14 1e
USER 6003 INFO:
a, 14, 1e
USER 6012 INFO:
a141e
USER 6013 INFO:
a, 14, 1e
USER 6015 INFO:
hello
USER 6016 INFO:
h e l l o 
USER 6017 INFO:
104, 101, 108, 108, 111, 0
USER 6021 INFO:
"{1, 2} {10, 20}"
USER 6022 INFO:
"{1, 2}, {10, 20}"
USER 6023 INFO:
"{1, 2}{10, 20}"
USER 6024 INFO:
"{1, 2}, {10, 20}"
USER 6025 INFO:
"{1, 2}, {10, 20}"
USER 6026 INFO:
"{1, 2} {10, 20}"
USER 6031 INFO:
10
20
30
USER 6032 INFO:
10:20:30
USER 6033 INFO:
10	20	30
USER 6034 INFO:
10.20.30
